gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -0,0 +1,158 @@
+
+        self._asset_volume_time_series_buffer: Dict[DateTime, Dict] = {}
+
+    def add(self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add asset time series to asset volume time series."""
+
+        self._add_total_energy_bought(asset_stats, market_type)
+        self._add_total_energy_sold(asset_stats, market_type)
+
+    def save_time_series(self):
+        """Save asset volume time series in the DB."""
+        # TODO: should be implemented by Kamil.
+        raise NotImplementedError()
+
+    def _add_total_energy_bought(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of bought energy."""
+        energy_kWh = asset_stats.total_energy_bought
+        if energy_kWh <= 0:
+            return
+        profile = ForwardTradeProfileGenerator(self.asset_capacity).generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type)
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "bought_kWh")
+
+    def _add_total_energy_sold(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of sold energy."""
+        energy_kWh = asset_stats.total_energy_sold
+        if energy_kWh <= 0:
+            return
+        profile = ForwardTradeProfileGenerator(self.asset_capacity).generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type
+        )
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "sold_kWh")
+
+    def _add_to_volume_time_series(
+            self, time_slot: DateTime, value: float,
+            market_type: AvailableMarketTypes, attribute_name: str):
+        """Add time slot value to the correct time slot to asset volume time series."""
+        time_slot = self._adapt_time_slot(time_slot)
+        volume_time_series = self._get_asset_volume_time_series(time_slot.start_of("year"))
+        volume_time_series[time_slot][str(market_type)][attribute_name] += value
+
+    def _adapt_time_slot(self, time_slot: DateTime) -> DateTime:
+        """Find the containing time slot of the smaller time slot."""
+        if self.resolution == AggregationResolution.RES_15_MINUTES:
+            return time_slot.set(minute=(time_slot.minute // 15) * 15)
+        return time_slot.start_of(START_OF[self.resolution])
+
+    def _get_asset_volume_time_series(self, time_slot):
+        """Return asset volume time series for the required time slot.
+        If not found in the buffer, it tries fetching it from DB.
+        If not found in the DB, it will generate a new one for the whole year."""
+        if time_slot in self._asset_volume_time_series_buffer:
+            time_series = self._asset_volume_time_series_buffer[time_slot]
```

Comment:
That's true.

Reply:
Is this really needed? This nesting will not be represented in the DB anyway right?

Reply:
It will not be represented in the DB like this, but it's needed here. Let me explain more ...

Consider you want to update the graphs with a new trade just received from the weekly product.
The timeslot for the weekly product is: `2019-12-30T00:00`. Part of the sold energy for the weekly product is in 2019 and the rest of it is in 2020.

Now consider we want the monthly resolution of the device volume time series which starts from 1s of Jan; should we add the energy of the weekly product to Jan 2020? Should we add it to December 2019?

IMO, we should split the weekly product into smaller time quantoms and then distributed each quantom regarding to its year.

And for this purpose, we need volume time series of both 2019 and 2020. So we fetch them both (on demand) and save them after updating them.

Reply:
IMO adding the week that starts at `2019-12-30T00:00` to 2019 is sufficient. No need to complicate the code only for one corner case. But we need to discuss this in the team.

Reply:
Calculating for the whole year actually makes the code look simpler.
If we don't do that, then when other 2019 weeks come to get added, we have to recalculate time series for 2019 to compensate for the missing weeks (and this can make code even more complex).